ksqldb-engine/src/main/java/io/confluent/ksql/format/DefaultFormatInjector.java

```diff
@@ -0,0 +1,178 @@
+
+import io.confluent.ksql.parser.KsqlParser.PreparedStatement;
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.exception.ParseFailedException;
+import io.confluent.ksql.parser.properties.with.CreateSourceProperties;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.CreateSource;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.properties.with.CommonCreateConfigs;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import io.confluent.ksql.statement.Injector;
+import io.confluent.ksql.util.ErrorMessageUtil;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.KsqlStatementException;
+import java.util.Optional;
+
+/**
+ * An injector which injects the key and value formats into the supplied {@code statement}.
+ *
+ * <p>The key format is only injected if:
+ * <ul>
+ * <li>The statement is a CT/CS.</li>
+ * <li>The statement does not specify the FORMAT property in its WITH clause.</li>
+ * <li>The statement does not specify the KEY_FORMAT property in its WITH clause.</li>
+ * </ul>
+ *
+ * <p>Similarly, the value format is only injected if the above conditions are met,
+ * where the KEY_FORMAT property is replaced with the VALUE_FORMAT property accordingly.
+ *
+ * <p>If any of the above are not true then the {@code statement} is returned unchanged.
+ *
+ * <p>This injector also validates that the KEY_FORMAT property is only supplied if the
+ * relevant feature flag is enabled.
+ */
+public class DefaultFormatInjector implements Injector {
+
+  public DefaultFormatInjector() {
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> ConfiguredStatement<T> inject(
+      final ConfiguredStatement<T> statement
+  ) {
+    if (statement.getStatement() instanceof CreateSource) {
+      final ConfiguredStatement<CreateSource> createStatement =
+          (ConfiguredStatement<CreateSource>) statement;
+
+      if (createStatement.getStatement().getProperties().getKeyFormatName().isPresent()) {
+        throwIfKeyFormatDisabled(createStatement);
+      }
+
+      try {
+        return (ConfiguredStatement<T>) forCreateStatement(createStatement).orElse(createStatement);
+      } catch (final KsqlStatementException e) {
+        throw e;
+      } catch (final KsqlException e) {
+        throw new KsqlStatementException(
+            ErrorMessageUtil.buildErrorMessage(e),
+            statement.getStatementText(),
+            e.getCause());
+      }
+    }
+
+    if (statement.getStatement() instanceof CreateAsSelect) {
+      final ConfiguredStatement<CreateAsSelect> createAsSelect =
+          (ConfiguredStatement<CreateAsSelect>) statement;
+
+      if (createAsSelect.getStatement().getProperties().getKeyFormatInfo().isPresent()) {
+        throwIfKeyFormatDisabled(createAsSelect);
+      }
+    }
+
+    return statement;
+  }
```

Comment:
something like:

```java
@SupressWarning({"unchecked", "rawType"})
private <T extends Statement> ConfiguredStatement<T> handleCreateSource(
    final ConfiguredStatement<CreateSource> statement
  ) {

  ...

  // Safe to cast as we _know_ `T` is `CreateSource`.
  return (ConfiguredStatement)result;
}
```

Reply:
Done, except there's still an unchecked cast warning on the main `inject()` method since checking that `statement.getStatement() instanceof CreateSource` isn't enough to guarantee a clean cast for `(ConfiguredStatement<CreateSource>) statement`. Am I missing something?

Reply:
I guess you could do:

```java
@SupressWarning({"unchecked", "rawType"})
private <T extends Statement> ConfiguredStatement<T> handleCreateSource(
    final ConfiguredStatement<T> statement
  ) {
  if (!statement.getStatement() instance of CreateSource) {
     throw new IllegalArgumentException(something);
  }
  ...

  return (ConfiguredStatement)result;
}
```